kale/tests/unit_tests/test_metadata_utils.py

```diff
@@ -37,23 +38,64 @@
      }])
 ])
 def test_validate_volumes_metadata(volumes, target):
-    """Tests that volumes are correctly converted from list into dict
+    """Tests that volumes are correctly converted from list into dict."""
+    assert target == metadata_utils._validate_volumes_metadata(volumes)
 
-    """
-    assert target == _validate_volumes_metadata(volumes)
 
-
-@pytest.mark.parametrize("volumes", [
-    {
+@pytest.mark.parametrize("volumes,match", [
+    ([{
         'annotations': [{'key': 'a1', 'value': 'v1'}],
         'size': 5,
         'type': 'pv',
         'mount_point': '/'
-    }
+    }], "Volume spec: missing name value"),
+    ([{
+        'name': 'v1',
+        'annotations': [{'wronkey': 'a1', 'value': 'v1'}],
+        'size': 5,
+        'type': 'pv',
+        'mount_point': '/'
+    }], "Volume spec: volume annotations must be a "
+        "list of {'key': k, 'value': v} dicts"),
+    ([{
+        'name': 'v1',
+        'annotations': {'key': 'a1', 'value': 'v1'},
+        'size': 5,
+        'type': 'pv',
+        'mount_point': '/'
+    }], "Volume spec: annotations must be a list"),
 ])
-def test_convert_volume_annotations_exc(volumes):
-    """Tests that volumes are correctly converted from list into dict
+def test_convert_volume_annotations_exc(volumes, match):
+    """Tests that volumes are correctly converted from list into dict."""
+    with pytest.raises(ValueError, match=match):
+        metadata_utils._validate_volumes_metadata(volumes)
+
+
+@pytest.mark.parametrize("metadata,target", [
+    ({'pipeline_name': 'test',
+      'experiment_name': 'test'},
+     metadata_utils.DEFAULT_METADATA),
```

Comment:
I think we should `copy.deepcopy()` even that, because it should affect the next tests, depending on memory behavior.

Also, what's the purpose of two identical test parameters?

Reply:
Thanks for the `deepcopy` catch. As for the two identical test, this is an oversight of mine. This tests should have many more parameters, to test many more test cases. My intention was to set up the basic test and address the others in the future.